test/extensions/quic_listeners/quiche/integration/quic_http_integration_test.cc

```diff
@@ -362,12 +368,21 @@ TEST_P(QuicHttpIntegrationTest, Retry) { testRetry(); }
 
 TEST_P(QuicHttpIntegrationTest, UpstreamReadDisabledOnGiantResponseBody) {
   config_helper_.setBufferLimits(/*upstream_buffer_limit=*/1024, /*downstream_buffer_limit=*/1024);
-  testRouterRequestAndResponseWithBody(/*request_size=*/512, /*response_size=*/1024 * 1024, false);
+  testRouterRequestAndResponseWithBody(/*request_size=*/512, /*response_size=*/10 * 1024 * 1024,
+                                       false);
 }
 
 TEST_P(QuicHttpIntegrationTest, DownstreamReadDisabledOnGiantPost) {
   config_helper_.setBufferLimits(/*upstream_buffer_limit=*/1024, /*downstream_buffer_limit=*/1024);
-  testRouterRequestAndResponseWithBody(/*request_size=*/1024 * 1024, /*response_size=*/1024, false);
+  testRouterRequestAndResponseWithBody(/*request_size=*/10 * 1024 * 1024, /*response_size=*/1024,
+                                       false);
+}
+
+TEST_P(QuicHttpIntegrationTest, LargeFlowControlOnAndGiantBody) {
+  config_helper_.setBufferLimits(/*upstream_buffer_limit=*/128 * 1024,
+                                 /*downstream_buffer_limit=*/128 * 1024);
+  testRouterRequestAndResponseWithBody(/*request_size=*/10 * 1024 * 1024,
```

Comment:
Awesome!

WDYT of adding a unit test for the new scoped class working as intended?